### PR TITLE
ADBDEV-3852-3 Fix ORCA invalid processing of nested SubLinks under the aggregates.

### PR DESCRIPTION
**Pull request by bimboterminator1**
The ORCA optimizer could produce an invalid query plan and fallback to the
legacy optimizer due to incorrect processing of nested SubLinks (SubLink
contains one more SubLink or rtable subquery inside) during query normalization.

The issue could arise when query had a AggRef inside a Sublink and that
AggRef contained one more SubLink inside (see the example of the
query in the tests section).

While mutating the SubLink inside RunExtractAggregatesMutator, the Aggref branch
is executed for SubLink's query targetList, which mutates args of the AggRef.
And if inside this AggRef arg some Var had varlevelsup value greater than the
m_agg_levels_up, i.e the Var referenced a relation that is higher than the
AggRef, the Var was not mutated at all. The varlevelsup field was not modified
and because of that, when AggRef was pulled up to the zero query level (by
appending it to the context.m_lower_table_tlist), the Var, whose varlevelsup
was higher than AggRef's, and which remained unchanged, started referencing
the wrong relation. This could lead to the fallback or invalid plan construction
when processing attributes at further planning stages.

This patch makes the RunExtractAggregatesMutator modify the varlevelsup of the
Var relatively the AggRef level in order to preserve proper query level.

### CODE REVIEW

File: src/backend/gpopt/translate/CQueryMutators.cpp [hunk 1]
```diff
@@ -702,7 +702,7 @@ CQueryMutators::RunExtractAggregatesMutator(Node *node,
 				//
 				// Note the foo.a var which is in sum() in a subquery must now become a
 				// var referencing the current query level.
-				var->varlevelsup = 0;
+				var->varlevelsup -= context->m_agg_levels_up;
```
Comment (andr-sokolov) [inline src/backend/gpopt/translate/CQueryMutators.cpp:709]: The comment above is about the previous code. Rewrite please the comment to describe more general case

**Comment by andr-sokolov** [comment]
"And if a Var had varlevelsup value greater than the agglevelsup" - may be "m_agg_levels_up"?

File: src/backend/gpopt/translate/CQueryMutators.cpp (hunk 1 above)
Comment (bimboterminator1) [reply src/backend/gpopt/translate/CQueryMutators.cpp:709]: done

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/52684

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/52818

File: src/backend/gpopt/translate/CQueryMutators.cpp (hunk 1 above)
Comment (andr-sokolov) [reply src/backend/gpopt/translate/CQueryMutators.cpp:709]: The comment should be shorter. I suggest to remove the first example. The second example is more general and this one example is enough to understand the algorithm

Comment (bimboterminator1) [reply src/backend/gpopt/translate/CQueryMutators.cpp:709]: Evaluate, please 
```
If Var references the top level query inside an Aggref that also
references top level query, the Aggref is moved to the derived query
(see comments in Aggref if-case above). Thus, these Vars
are brought up together with the Aggref, i.e.
their varlevelsup is decreased by m_agg_levels_up.
e.g:
select (select max(select foo.a)) from foo;
is transformed into
select (select fnew.max_t)
from (select max(select foo.a) max_t from foo) fnew;
Here the foo.a inside max referenced top level RTE foo at
varlevelsup = 2 inside the Aggref at agglevelsup 1. Then the
Aggref is brought up to the top-query-level of fnew and foo.a
inside Aggref is decreased by original Aggref's level.
```

Comment (andr-sokolov) [reply src/backend/gpopt/translate/CQueryMutators.cpp:709]: "If Var references the top level query inside an Aggref that also references top level query". The condition on 688th line checks only Var's levels up

Comment (bimboterminator1) [reply src/backend/gpopt/translate/CQueryMutators.cpp:709]:
```
We previously started to mutate the Aggref, that references the top level query.
This Aggref is going to be moved to the derived query (see comments in Aggref if-case above). 
Therefore, if we are mutating Vars inside the Aggref, and these Vars reference the
top level query (varlevelsup = m_current_query_level) as well, we must change
their varlevelsup field in order to preserve correct reference level. i.e these Vars
are pulled up as the part of Aggref by the m_agg_levels_up.
e.g:
select (select max((select foo.a))) from foo;
is transformed into
select (select fnew.max_t)
from (select max((select foo.a)) max_t from foo) fnew;
Here the foo.a inside max referenced top level RTE foo at
varlevelsup = 2 inside the Aggref at agglevelsup 1. Then the
Aggref is brought up to the top-query-level of fnew and foo.a
inside Aggref is decreased by original Aggref's level.
```

**Comment by BenderArenadata** [comment]
Allure report

**Review by andr-sokolov — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/53049

**Pull request merged**